streaming/base/shared.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2022 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Synchronization primitives that live in shared memory.
+
+For when using `threading` or `multiprocessing` from the python standard library won't do, because
+we are coordinating separately instantiated pytorch worker processes.
+"""
+
+import os
+from multiprocessing.shared_memory import SharedMemory
+from time import sleep
+
+import numpy as np
+from filelock import FileLock
+
+TICK = 0.07
+
+
+class SharedBarrier:
+    """A barrier that works inter-process using a file lock and shared memory.
+
+    We set the number of processes (and thereby initialilze num_exit) on the first time this object
+    is called. This is because the object is created in a per-rank process, and called by worker
+    processes.
+
+    Args:
+        filelock_path (str): Path to lock file on local filesystem.
+        shm_path (str): Shared memory object name in /dev/shm.
+    """
+
+    def __init__(self, filelock_path: str, shm_path: str) -> None:
+        self.filelock_path = filelock_path
+        self.shm_path = shm_path
+
+        # Create filelock.
+        dirname = os.path.dirname(filelock_path)
+        os.makedirs(dirname, exist_ok=True)
+        self.lock = FileLock(filelock_path)
+
+        # Create three int32 fields in shared memory: num_enter, num_exit, flag.
+        size = 3 * np.int32(0).nbytes
+        try:
+            self._shm = SharedMemory(shm_path, True, size)
+        except FileExistsError:
+            self._shm = SharedMemory(shm_path, False, size)
+        self._arr = np.ndarray(3, buffer=self._shm.buf, dtype=np.int32)
+        self.num_enter = 0
+        self.num_exit = -1
+        self.flag = True
+
+    def __del__(self):
+        """Destructor clears array that references shm."""
+        try:
+            del self._arr
+            self._shm.close()
```

Comment:
I added the `close()` and `unlink()` method since looks like there was a memory link. I was getting a below warning message while working on a unit test

```
/opt/homebrew/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked shared_memory objects to clean up at shutdown
```

Ref: https://docs.python.org/3/library/multiprocessing.shared_memory.html#multiprocessing.shared_memory.SharedMemory.close